File: packages/sdk/src/cost-model.ts
/**
 * Cost-Aware Verification Model
 *
 * TypeScript implementation of gas estimation and verification path selection.
 * Provides pre-verification cost analysis for universal ZK verification.
 *
 * Features:
 * - Pre-verification gas estimation - Know costs before committing
 * - Cost comparison across proof systems
 * - Budget-aware verification routing
 * - Batch verification discount calculation
 *
 * @see packages/stylus/src/cost_model.rs for Rust implementation
 */

import { ProofType } from "./types";
import { UniversalProofDescriptor } from "./upd";

/**
 * Cost breakdown as percentages
 */
export interface CostBreakdown {
  /** Base cost percentage */
  basePct: number;
  /** Per-input cost percentage */
  perInputPct: number;
  /** Per-byte cost percentage */
  perBytePct: number;
}

/**
 * Cost comparison result
 */
export enum CostComparison {
  /** First is cheaper */
  FirstCheaper = "FirstCheaper",
  /** Second is cheaper */
  SecondCheaper = "SecondCheaper",
  /** Costs are equal */
  Equal = "Equal",
}

/**
 * Gas limit recommendations
 */
export interface GasLimitRecommendation {
  /** Minimum recommended gas limit */
  minimum: bigint;
  /** Comfortable margin (10% buffer) */
  recommended: bigint;
  /** Safe maximum for complex operations */
  maximum: bigint;
}

/**
 * Normalized verification cost with breakdown
 *
 * Captures the full cost structure for a verification operation,
 * enabling both estimation and comparison across proof systems.
 */
export class VerificationCost {
  /** Proof system type */
  public readonly proofType: ProofType;

  /** Base gas cost */
  public readonly baseCost: bigint;

  /** Per-input gas cost */
  public readonly perInputCost: bigint;

  /** Per-byte gas cost */
  public readonly perByteCost: bigint;

  /** Number of public inputs */
  public readonly publicInputs: number;

  /** Proof size in bytes */
  public readonly proofBytes: number;

  /** Estimated total gas */
  public readonly estimatedTotal: bigint;

  private constructor(params: {
    proofType: ProofType;
    baseCost: bigint;
    perInputCost: bigint;
    perByteCost: bigint;
    publicInputs: number;
    proofBytes: number;
  }) {
    this.proofType = params.proofType;
    this.baseCost = params.baseCost;
    this.perInputCost = params.perInputCost;
    this.perByteCost = params.perByteCost;
    this.publicInputs = params.publicInputs;
    this.proofBytes = params.proofBytes;

    // Calculate total
    this.estimatedTotal =
      this.baseCost +
      this.perInputCost * BigInt(this.publicInputs) +
      this.perByteCost * BigInt(this.proofBytes);
  }

  /**
   * Calculate verification cost for Groth16 proof
   *
   * Cost model:
   * - Base: 250,000 gas (4 pairing checks)
   * - Per input: 40,000 gas (MSM operations)
   * - Per byte: 0 (fixed-size proof)
   *
   * Typical total: 250k + 40k * inputs ≈ 330k for 2 inputs
   */
  static forGroth16(publicInputs: number): VerificationCost {
    return new VerificationCost({
      proofType: ProofType.Groth16,
      baseCost: 250_000n,
      perInputCost: 40_000n,
      perByteCost: 0n,
      publicInputs,
      proofBytes: 256, // Standard Groth16 proof size
    });
  }

  /**
   * Calculate verification cost for PLONK proof
   *
   * Cost model:
   * - Base: 350,000 gas (more pairings than Groth16)
   * - Per input: 10,000 gas (commitment lookups)
   * - Per byte: 0 (fixed-size proof)
   *
   * Typical total: 350k + 10k * inputs ≈ 390k for 4 inputs
   */
  static forPlonk(
    publicInputs: number,
    circuitSize: number = 0,
  ): VerificationCost {
    // Circuit size adds a small overhead (not modeled precisely)
    const circuitOverhead = BigInt(Math.floor(circuitSize / 10000)) * 1000n;

    return new VerificationCost({
      proofType: ProofType.PLONK,
      baseCost: 350_000n + circuitOverhead,
      perInputCost: 10_000n,
      perByteCost: 0n,
      publicInputs,
      proofBytes: 800, // Typical PLONK proof size
    });
  }

  /**
   * Calculate verification cost for STARK proof
   *
   * Cost model:
   * - Base: 200,000 gas (hash operations)
   * - Per input: 5,000 gas (minimal)
   * - Per byte: 10 gas (FRI layer verification)
   *
   * Typical total: 200k + 10 * proof_bytes ≈ 700k for 50KB proof
   */
  static forStark(
    traceLength: number,
    securityBits: number = 100,
  ): VerificationCost {
    // Proof size is approximately log2(traceLength) * securityBits * 32
    const layers = Math.ceil(Math.log2(traceLength || 1));
    const proofBytes = layers * securityBits * 32;

    return new VerificationCost({
      proofType: ProofType.STARK,
      baseCost: 200_000n,
      perInputCost: 5_000n,
      perByteCost: 10n,
      publicInputs: 0, // STARK typically has minimal public inputs
      proofBytes,
    });
  }

  /**
   * Create cost from UniversalProofDescriptor
   *
   * Uses the descriptor's metadata to compute accurate gas estimate.
   */
  static fromDescriptor(
    descriptor: UniversalProofDescriptor,
  ): VerificationCost {
    const proofType = descriptor.proofType();

    if (proofType === null) {
      throw new Error("Unknown proof type in descriptor");
    }

    switch (proofType) {
      case ProofType.Groth16:
        return new VerificationCost({
          proofType: ProofType.Groth16,
          baseCost: 250_000n,
          perInputCost: 40_000n,
          perByteCost: 0n,
          publicInputs: descriptor.publicInputCount,
          proofBytes: descriptor.proofLength,
        });

      case ProofType.PLONK:
        return new VerificationCost({
          proofType: ProofType.PLONK,
          baseCost: 350_000n,
          perInputCost: 10_000n,
          perByteCost: 0n,
          publicInputs: descriptor.publicInputCount,
          proofBytes: descriptor.proofLength,
        });

      case ProofType.STARK:
        return new VerificationCost({
          proofType: ProofType.STARK,
          baseCost: 200_000n,
          perInputCost: 5_000n,
          perByteCost: 10n,
          publicInputs: descriptor.publicInputCount,
          proofBytes: descriptor.proofLength,
        });

      default:
        throw new Error(`Unsupported proof type: ${proofType}`);
    }
  }

  /**
   * Check if this cost is cheaper than another
   */
  cheaperThan(other: VerificationCost): boolean {
    return this.estimatedTotal < other.estimatedTotal;
  }

  /**
   * Get the gas savings compared to another cost
   * Positive means this is cheaper, negative means other is cheaper
   */
  savingsVs(other: VerificationCost): bigint {
    return other.estimatedTotal - this.estimatedTotal;
  }

  /**
   * Check if cost is within a gas budget
   */
  withinBudget(gasLimit: bigint): boolean {
    return this.estimatedTotal <= gasLimit;
  }

  /**
   * Get cost breakdown as percentages
   */
  costBreakdown(): CostBreakdown {
    if (this.estimatedTotal === 0n) {
      return { basePct: 0, perInputPct: 0, perBytePct: 0 };
    }

    const total = Number(this.estimatedTotal);
    const inputTotal = Number(this.perInputCost) * this.publicInputs;
    const byteTotal = Number(this.perByteCost) * this.proofBytes;

    return {
      basePct: (Number(this.baseCost) / total) * 100,
      perInputPct: (inputTotal / total) * 100,
      perBytePct: (byteTotal / total) * 100,
    };
  }
}

/**
 * Compare verification costs and return recommendation
 */
export function compareCosts(
  a: VerificationCost,
  b: VerificationCost,
): CostComparison {
  if (a.estimatedTotal < b.estimatedTotal) {
    return CostComparison.FirstCheaper;
  } else if (a.estimatedTotal > b.estimatedTotal) {
    return CostComparison.SecondCheaper;
  } else {
    return CostComparison.Equal;
  }
}

/**
 * Select cheapest verification path from multiple options
 *
 * @param costs Array of verification costs
 * @returns Index of the cheapest option, or null if empty
 */
export function selectCheapest(costs: VerificationCost[]): number | null {
  if (costs.length === 0) {
    return null;
  }

  let cheapestIndex = 0;
  let cheapestCost = costs[0].estimatedTotal;

  for (let i = 1; i < costs.length; i++) {
    if (costs[i].estimatedTotal < cheapestCost) {
      cheapestCost = costs[i].estimatedTotal;
      cheapestIndex = i;
    }
  }

  return cheapestIndex;
}

/**
 * Budget-aware verification routing
 *
 * Returns true if verification should proceed, false if too expensive.
 *
 * @param cost Verification cost
 * @param gasLimit Available gas limit
 * @param safetyMarginPct Safety margin percentage (0-100)
 */
export function shouldVerify(
  cost: VerificationCost,
  gasLimit: bigint,
  safetyMarginPct: number = 10,
): boolean {
  // Calculate effective limit with safety margin
  const margin = (gasLimit * BigInt(safetyMarginPct)) / 100n;
  const effectiveLimit = gasLimit - margin;

  return cost.estimatedTotal <= effectiveLimit;
}

/**
 * Estimate total cost for batch verification
 *
 * Batch verification can be more efficient due to shared setup costs.
 * Discount factor: 5% per additional proof (up to 30% max discount)
 *
 * @param costs Array of individual verification costs
 * @returns Total batch cost after discounts
 */
export function estimateBatchCost(costs: VerificationCost[]): bigint {
  if (costs.length === 0) {
    return 0n;
  }

  // Sum all individual costs
  let total = 0n;
  for (const cost of costs) {
    total += cost.estimatedTotal;
  }

  // Apply batch discount (5% per additional proof, max 30%)
  const discountPct = Math.min((costs.length - 1) * 5, 30);
  const discount = (total * BigInt(discountPct)) / 100n;

  return total - discount;
}

/**
 * Get gas limit recommendation for a proof type
 */
export function getGasLimitRecommendation(
  proofType: ProofType,
  publicInputs: number,
): GasLimitRecommendation {
  let cost: VerificationCost;

  switch (proofType) {
    case ProofType.Groth16:
      cost = VerificationCost.forGroth16(publicInputs);
      break;
    case ProofType.PLONK:
      cost = VerificationCost.forPlonk(publicInputs);
      break;
    case ProofType.STARK:
      cost = VerificationCost.forStark(1024, 100); // Default trace length
      break;
    default:
      cost = VerificationCost.forGroth16(publicInputs);
  }

  const minimum = cost.estimatedTotal;
  const recommended = (minimum * 110n) / 100n; // 10% buffer
  const maximum = (minimum * 150n) / 100n; // 50% buffer

  return { minimum, recommended, maximum };
}

File: packages/stylus/src/cost_model.rs
//! Cost-Aware Verification Model
//!
//! Provides gas estimation and verification path selection for universal ZK verification.
//!
//! # Features
//! - **Pre-verification gas estimation** - Know costs before committing
//! - **Cost comparison** - Select cheapest proof system for equivalent security
//! - **Budget-aware routing** - Reject proofs that exceed gas limits
//!
//! # Gas Cost Models
//!
//! | System | Base | Per Input | Per Byte | Notes |
//! |--------|------|-----------|----------|-------|
//! | Groth16 | 250k | 40k | 0 | Pairing-based, fixed size |
//! | PLONK | 350k | 10k | 0 | More pairings |
//! | STARK | 200k | 5k | 10 | FRI verification |

extern crate alloc;

use crate::types::{ProofType, UniversalProofDescriptor};
use crate::verifier_traits::GasCost;

/// Normalized verification cost with breakdown
///
/// Captures the full cost structure for a verification operation,
/// enabling both estimation and comparison across proof systems.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct VerificationCost {
    /// Base gas cost (constant overhead)
    pub base_gas: u64,

    /// Per-public-input gas cost component
    pub per_input_gas: u64,

    /// Per-proof-byte gas cost component
    pub per_byte_gas: u64,

    /// Number of public inputs
    pub public_input_count: usize,

    /// Proof size in bytes
    pub proof_size: usize,

    /// Computed total gas cost
    pub estimated_total: u64,

    /// Proof system this cost is for
    pub proof_system: ProofType,
}

impl VerificationCost {
    /// Calculate verification cost for Groth16 proof
    ///
    /// Cost model:
    /// - Base: 250,000 gas (4 pairing checks)
    /// - Per input: 40,000 gas (MSM operations)
    /// - Per byte: 0 (fixed-size proof)
    ///
    /// Typical total: 250k + 40k * inputs ≈ 330k for 2 inputs
    pub fn for_groth16(public_inputs: usize) -> Self {
        let model = GasCost::groth16();
        let proof_size = 256; // Standard Groth16 proof size

        let estimated_total = model.estimate(public_inputs, proof_size);

        Self {
            base_gas: model.base,
            per_input_gas: model.per_public_input,
            per_byte_gas: model.per_proof_byte,
            public_input_count: public_inputs,
            proof_size,
            estimated_total,
            proof_system: ProofType::Groth16,
        }
    }

    /// Calculate verification cost for PLONK proof
    ///
    /// Cost model:
    /// - Base: 350,000 gas (more pairings than Groth16)
    /// - Per input: 10,000 gas (commitment lookups)
    /// - Per byte: 0 (fixed-size proof)
    ///
    /// Typical total: 350k + 10k * inputs ≈ 390k for 4 inputs
    pub fn for_plonk(public_inputs: usize, circuit_size: usize) -> Self {
        let model = GasCost::plonk();

        // PLONK proof size varies with circuit size (roughly log2(n) commitments)
        let proof_size = 800 + (circuit_size.ilog2() as usize * 64);

        let estimated_total = model.estimate(public_inputs, proof_size);

        Self {
            base_gas: model.base,
            per_input_gas: model.per_public_input,
            per_byte_gas: model.per_proof_byte,
            public_input_count: public_inputs,
            proof_size,
            estimated_total,
            proof_system: ProofType::PLONK,
        }
    }

    /// Calculate verification cost for STARK proof
    ///
    /// Cost model:
    /// - Base: 200,000 gas (hash operations)
    /// - Per input: 5,000 gas (minimal)
    /// - Per byte: 10 gas (FRI layer verification)
    ///
    /// Typical total: 200k + 10 * proof_bytes ≈ 700k for 50KB proof
    pub fn for_stark(trace_length: usize, security_bits: u8) -> Self {
        let model = GasCost::stark();

        // STARK proof size: ~40KB base + scaling with trace length and security
        // Rough formula: base + log2(trace_length) * security_bits * factor
        let log_trace = trace_length.max(64).ilog2() as usize;
        let security_factor = security_bits as usize / 32;
        let proof_size = 40_000 + (log_trace * security_factor * 2_000);

        let estimated_total = model.estimate(4, proof_size); // STARKs typically have few public inputs

        Self {
            base_gas: model.base,
            per_input_gas: model.per_public_input,
            per_byte_gas: model.per_proof_byte,
            public_input_count: 4,
            proof_size,
            estimated_total,
            proof_system: ProofType::STARK,
        }
    }

    /// Create cost from UniversalProofDescriptor
    ///
    /// Uses the descriptor's metadata to compute accurate gas estimate.
    pub fn from_descriptor(descriptor: &UniversalProofDescriptor) -> Self {
        let model = match descriptor.proof_system_id {
            0 => GasCost::groth16(),
            1 => GasCost::plonk(),
            2 => GasCost::stark(),
            _ => GasCost::groth16(), // Fallback
        };

        let proof_type = match descriptor.proof_system_id {
            0 => ProofType::Groth16,
            1 => ProofType::PLONK,
            2 => ProofType::STARK,
            _ => ProofType::Groth16,
        };

        let estimated_total = model.estimate(
            descriptor.public_input_count as usize,
            descriptor.proof_length as usize,
        );

        Self {
            base_gas: model.base,
            per_input_gas: model.per_public_input,
            per_byte_gas: model.per_proof_byte,
            public_input_count: descriptor.public_input_count as usize,
            proof_size: descriptor.proof_length as usize,
            estimated_total,
            proof_system: proof_type,
        }
    }

    /// Check if this cost is cheaper than another
    pub fn cheaper_than(&self, other: &Self) -> bool {
        self.estimated_total < other.estimated_total
    }

    /// Get the gas savings compared to another cost
    pub fn savings_vs(&self, other: &Self) -> i64 {
        other.estimated_total as i64 - self.estimated_total as i64
    }

    /// Check if cost is within a gas budget
    pub fn within_budget(&self, gas_limit: u64) -> bool {
        self.estimated_total <= gas_limit
    }

    /// Get cost breakdown as percentage
    pub fn cost_breakdown(&self) -> CostBreakdown {
        let input_cost = self.per_input_gas * self.public_input_count as u64;
        let byte_cost = self.per_byte_gas * self.proof_size as u64;

        CostBreakdown {
            base_pct: (self.base_gas * 100) / self.estimated_total.max(1),
            input_pct: (input_cost * 100) / self.estimated_total.max(1),
            byte_pct: (byte_cost * 100) / self.estimated_total.max(1),
        }
    }
}

/// Cost breakdown as percentages
#[derive(Debug, Clone, Copy)]
pub struct CostBreakdown {
    /// Percentage of cost from base gas
    pub base_pct: u64,
    /// Percentage of cost from public inputs
    pub input_pct: u64,
    /// Percentage of cost from proof bytes
    pub byte_pct: u64,
}

/// Cost comparison result
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CostComparison {
    /// First is cheaper
    FirstCheaper,
    /// Second is cheaper
    SecondCheaper,
    /// Costs are equal
    Equal,
}

/// Compare verification costs and return recommendation
pub fn compare_costs(a: &VerificationCost, b: &VerificationCost) -> CostComparison {
    match a.estimated_total.cmp(&b.estimated_total) {
        core::cmp::Ordering::Less => CostComparison::FirstCheaper,
        core::cmp::Ordering::Greater => CostComparison::SecondCheaper,
        core::cmp::Ordering::Equal => CostComparison::Equal,
    }
}

/// Select cheapest verification path from multiple options
///
/// Returns the index of the cheapest option, or None if empty.
pub fn select_cheapest(costs: &[VerificationCost]) -> Option<usize> {
    costs
        .iter()
        .enumerate()
        .min_by_key(|(_, cost)| cost.estimated_total)
        .map(|(idx, _)| idx)
}

/// Budget-aware verification routing
///
/// Returns true if verification should proceed, false if too expensive.
pub fn should_verify(cost: &VerificationCost, gas_limit: u64, safety_margin_pct: u8) -> bool {
    // Apply safety margin (e.g., 10% buffer for gas price fluctuations)
    let margin = (cost.estimated_total * safety_margin_pct as u64) / 100;
    let required_budget = cost.estimated_total + margin;

    required_budget <= gas_limit
}

/// Estimate total cost for batch verification
///
/// Batch verification can be more efficient due to shared setup costs.
/// Discount factor: 5% per additional proof (up to 30% max discount)
pub fn estimate_batch_cost(costs: &[VerificationCost]) -> u64 {
    if costs.is_empty() {
        return 0;
    }

    let base_total: u64 = costs.iter().map(|c| c.estimated_total).sum();

    // Calculate batch discount
    let discount_pct = core::cmp::min((costs.len() - 1) * 5, 30) as u64;
    let discount = (base_total * discount_pct) / 100;

    base_total - discount
}

/// Gas limit recommendations
#[derive(Debug, Clone, Copy)]
pub struct GasLimitRecommendation {
    /// Minimum gas for basic verification
    pub minimum: u64,
    /// Recommended gas with safety buffer
    pub recommended: u64,
    /// Maximum expected under worst case
    pub maximum: u64,
}

impl GasLimitRecommendation {
    /// Get gas limit recommendation for a proof type
    pub fn for_proof_type(proof_type: ProofType, public_inputs: usize) -> Self {
        let cost = match proof_type {
            ProofType::Groth16 => VerificationCost::for_groth16(public_inputs),
            ProofType::PLONK => VerificationCost::for_plonk(public_inputs, 65536),
            ProofType::STARK => VerificationCost::for_stark(1024, 128),
        };

        Self {
            minimum: cost.estimated_total,
            recommended: cost.estimated_total * 12 / 10, // +20% buffer
            maximum: cost.estimated_total * 15 / 10,     // +50% worst case
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloc::vec;

    #[test]
    fn test_groth16_cost() {
        let cost = VerificationCost::for_groth16(4);

        assert_eq!(cost.proof_system, ProofType::Groth16);
        assert!(cost.estimated_total > 180_000);
        assert!(cost.estimated_total < 300_000);
    }

    #[test]
    fn test_plonk_cost() {
        let cost = VerificationCost::for_plonk(4, 65536);

        assert_eq!(cost.proof_system, ProofType::PLONK);
        assert!(cost.estimated_total > 150_000);
    }

    #[test]
    fn test_stark_cost() {
        let cost = VerificationCost::for_stark(1024, 128);

        assert_eq!(cost.proof_system, ProofType::STARK);
        // STARK should be more expensive due to larger proofs
        assert!(cost.estimated_total > 400_000);
    }

    #[test]
    fn test_cost_comparison() {
        let groth16 = VerificationCost::for_groth16(4);
        let plonk = VerificationCost::for_plonk(4, 65536);

        // Groth16 (200k + 4*6.5k = 226k) vs PLONK (180k + 4*.5k = 182k)
        // PLONK is now cheaper for small inputs due to lower pairing count (2 vs 4) 
        // and cheaper inputs!?
        // Wait, Groth16 does 4 pairings. PLONK does 2 pairings.
        // Groth16 Base 200k. PLONK Base 180k (conservative).
        // Actually, PLONK with 2 pairings (113k) + MSMs might be CHEAPER than Groth16 constant 4 pairings (181k).
        // So PLONK should be cheaper!
        assert!(plonk.cheaper_than(&groth16));
        assert_eq!(compare_costs(&plonk, &groth16), CostComparison::FirstCheaper);
    }

    #[test]
    fn test_select_cheapest() {
        let options = vec![
            VerificationCost::for_stark(1024, 128),
            VerificationCost::for_groth16(4),
            VerificationCost::for_plonk(4, 65536),
        ];

        let cheapest = select_cheapest(&options);
        assert_eq!(cheapest, Some(2)); // PLONK is index 2, now cheaper (182k vs 226k)
    }

    #[test]
    fn test_budget_check() {
        let cost = VerificationCost::for_groth16(4);

        assert!(should_verify(&cost, 500_000, 10));
        assert!(!should_verify(&cost, 100_000, 10));
    }

    #[test]
    fn test_batch_discount() {
        let single = VerificationCost::for_groth16(4).estimated_total;
        let batch_3 = estimate_batch_cost(&[
            VerificationCost::for_groth16(4),
            VerificationCost::for_groth16(4),
            VerificationCost::for_groth16(4),
        ]);

        // 3 proofs should get 10% discount (2 * 5%)
        let expected = (single * 3) * 90 / 100;
        assert_eq!(batch_3, expected);
    }

    #[test]
    fn test_gas_recommendation() {
        let rec = GasLimitRecommendation::for_proof_type(ProofType::Groth16, 4);

        assert!(rec.minimum < rec.recommended);
        assert!(rec.recommended < rec.maximum);
    }
}
